docs/guide.md

```diff
@@ -184,6 +184,96 @@ This will check for the `?profile` query param on each request and if found,
 it starts profiling. After each request where the profiler was running it
 creates the html output and returns that instead of the actual response.
 
+### Profile a web request in FastAPI
+
+To profile call stacks in FastAPI, you can write a middleware extension for pyinstrument.
+
+The following class will instrument the code called for a web request and output text or HTML.
+
+```python
+import time
+
+from pyinstrument import Profiler
+
+
+class FastApiProfiler:
+    def __init__(
+        self,
+        interval: float = 0.0001,
+    ):
+        self._profiler = Profiler(interval=interval, async_mode="enabled")
+
+    def start(self) -> None:
+        self._profiler.start()
+        self._begin = time.perf_counter()
+
+    def stop(self):
+        self._profiler.stop()
+        self._end = time.perf_counter()
+
+    def get_result_html(self) -> str:
+        return self._profiler.output_html()
+
+    def get_result_text(self, color=True) -> str:
+        return self._profiler.output_text(color=color)
+
+```
+
+To enable this middleware, create an async function and decorate with `app.middleware('http')` where
+app is the name of your FastAPI application instance.
+
+Make sure you configure a setting to only make this available when required:
+
+```python
+PROFILING = True  # Set this from a settings model
+
+if PROFILING:
+    @app.middleware("http")
+    async def profile_request(request: Request, call_next):
+        profiling = request.query_params.get("profile", False)
+        if profiling:
+            profile = FastApiProfiler(interval=settings.profiling_interval)
+            profile.start()
+            await call_next(request)
+            profile.stop()
+            return HTMLResponse(profile.get_result_html())
+        else:
+            return await call_next(request)
```

Comment:
I wonder if the FastAPIProfiler class is needed at all? Any reason not to call straight to pyinstrument in this function?

```suggestion
    from pyinstrument import Profiler
    async def profile_request(request: Request, call_next):
        profiling = request.query_params.get("profile", False)
        if profiling:
            profiler = Profiler(interval=settings.profiling_interval, async_mode="enabled")
            profiler.start()
            await call_next(request)
            profiler.stop()
            return HTMLResponse(profiler.output_html())
        else:
            return await call_next(request)
```